content/en/blog/2022/cni-mode/index.md

```diff
@@ -0,0 +1,69 @@
+Second, in older Merbridge versions, only `connect` would process requests from the host, which was alright for same-node pod communication. However, in the case of cross-node pod communication, traffic will not be modified according to the previous logic. It means iptables are still necessary on the receiving end.
+<!--这次，我们依靠 XDP 程序，解决入口流量处理的问题。因为 XDP 程序需要挂载网卡，所以也需要借助 CNI。-->
+Here, we turned to the XDP program for ingress traffic processing. The XDP program needs to mount a network card, also making it necessary to have a CNI mode.
+<!--## CNI 如何解决问题？-->
+## How does CNI work
+<!--这里我们将探讨 CNI 的工作原理，以及如何使用 CNI 来解决问题。-->
+This section will explore how CNI works and how to use CNI to solve the issues mentioned above.
+<!--### 如何通过 CNI 让 eBPF 程序获取当前正在处理的 Pod IP？-->
+### How to use CNI to let eBPF have the current Pod IP?
+<!--我们通过 CNI，在 Pod 创建的时候，将 Pod 的 IP 信息写入一个 Map（`mark_pod_ips_map`），其 Key 为一个随机的值，Value 为 Pod 的 IP。然后，在当前 Pod 的 NetNS 里面监听一个特殊的端口 39807，将 Key 使用 `setsockopt` 写入这个端口 socket 的 mark。-->
+When a pod is created, we use CNI to write Pod IP into the map `mark_pod_ips_map` with a random value as the Key and the Pod IP as the value. Then, listen to a special port `39807` in the NetNS of the current Pod, and write the Key to the Mark of this port socket using `setsockopt`.
+<!--在 eBPF 中，我们通过 `bpf_sk_lookup_tcp` 取得端口 39807 的 Mark 信息，然后从 `mark_pod_ips_map` 中即可取得当前 NetNS（也是当期 Pod）的 IP。-->
+In eBPF, we get the Mark information of port `39807` through `bpf_sk_lookup_tcp`, and then we can get the current Pod IP, i.e., the corresponding NetNS from `mark_pod_ips_map`.
+<!--有了当前 Pod IP 之后，我们可以根据这个 Pod 的配置，确认流量处理路径（比如 `excludeOutboundPorts`）。-->
+With the current Pod IP, we can confirm the traffic processing path (such as `excludeOutboundPorts`) according to the configuration of this Pod.
+<!--同时，我们还使用 Pod 优化了之前解决四元组冲突的方案，改为使用 `bpf_bind` 绑定源 IP，目的 IP 直接使用 `127.0.0.1`，为了后续支持 IPv6 做准备。-->
+In addition, we also optimized the previous solution to quadruple conflicts by using `bpf_bind` to bind the source IP and using `127.0.0.1` as the destination IP, preparing for future support for IPv6.
+<!--### 如何处理入口流量？-->
+### How to handle ingress traffic?
+<!--为了能够处理入口流量，我们引入了 XDP 程序，XDP 程序作用在网卡上，能够对原始数据包做修改。-->
+In order to handle ingress traffic, we introduced the XDP program, which works on the network card and can modify the original data packets.
+<!--我们借助 XDP 程序，在流量到达 Pod 的时候，修改目的端口为 15006 以完成流量转发。-->
+We use the XDP program to modify the destination port as `15006` when the traffic reaches the Pod, so as to complete traffic forwarding.
+<!--同时考虑到可能存在主机直接访问 Pod 的情况，也为了减小影响范围，我们选择将 XDP 程序附加到 Pod 的网卡上。这就需要借助 CNI 的能力，在创建 Pod 时进行附加操作。-->
+Considering the possibility of direct host-pod communication, and in order to narrow down the affected scope, we attached the XDP program to the Pod's network card. This requires the ability of CNI to perform additional operations when creating Pods.
```

Comment:
Considering the possibility of direct host-pod communication, and in order to narrow down the affected scope, we attached the XDP program to the logical network card of a pod. This requires to use the capability of CNI to perform the `attach` action when creating a pod.